ctrl/ctrl_util.c

```diff
@@ -0,0 +1,217 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+
+#include "ctrl_util.h"
+#include "ctrl_caller.h"
+#include "ctrl.h"
+
+#include <event2/http.h>
+#include <event2/buffer.h>
+#include <event2/bufferevent.h>
+
+#include <string.h>
+
+#define MODULE_NAME "ctrl-utils"
+#define pv_log(level, msg, ...) vlog(MODULE_NAME, level, msg, ##__VA_ARGS__)
+#include "log.h"
+
+struct pv_ctrl_http_code_value {
+	int code;
+	const char *reason;
+};
+
+struct pv_ctrl_http_code_value http_codes[] = {
+	{ 507, "Insufficient Storage" },
+	{ 422, "Unprocessable Content" },
+	{ 409, "Conflict" },
+};
+
+#define PV_CTRL_UTILS_ERR_RSP "{\"Error\":\"%s\"}"
+
+static struct pv_ctrl_http_code_value ctrl_utils_get_http_value(int code)
+{
+	if (code > -1) {
+		pv_log(WARN, "pv_ctrl_codes should be < 0");
+		// this will cause a problem in the clint, a good clue
+		// to debug this
+		return (struct pv_ctrl_http_code_value){ -1, "" };
+	}
+	return http_codes[(-code) - 1];
+}
+
+void pv_ctrl_utils_send_ok(struct evhttp_request *req)
+{
+	evhttp_send_reply(req, HTTP_OK, NULL, NULL);
+}
+
+void pv_ctrl_utils_send_json(struct evhttp_request *req, int code,
+			     const char *reason, const char *json, ...)
+{
+	struct evbuffer *reply = evbuffer_new();
+	if (!reply) {
+		pv_log(DEBUG, "couldn't allocate reply buffer");
+		return;
+	}
+
+	va_list lst;
+	va_start(lst, json);
+
+	evbuffer_add_vprintf(reply, json, lst);
+	va_end(lst);
+
+	evhttp_add_header(evhttp_request_get_output_headers(req),
+			  "Content-Type", "application/json");
+	evhttp_send_reply(req, code, reason, reply);
+	evbuffer_free(reply);
+}
+
+void pv_ctrl_utils_send_error(struct evhttp_request *req, int code,
+			      const char *err_str)
+{
+	if (code > 0) {
+		pv_ctrl_utils_send_json(req, code, NULL, PV_CTRL_UTILS_ERR_RSP,
+					err_str);
+		return;
+	}
+
+	struct pv_ctrl_http_code_value v = ctrl_utils_get_http_value(code);
+	pv_ctrl_utils_send_json(req, v.code, v.reason, PV_CTRL_UTILS_ERR_RSP,
+				err_str);
+}
+
+int pv_ctrl_utils_split_path(const char *uri,
+			     char parts[PV_CTRL_MAX_SPLIT][NAME_MAX])
+{
+	if (!uri)
+		return 0;
+
+	const char *start = uri;
+	const char *end = uri;
+	int parts_count = 0;
+
+	while (*start) {
+		while (*start == '/')
+			start++;
+
+		if (!*start)
+			break;
+
+		end = start;
+		while (*end && *end != '/')
+			end++;
+
+		if (end > start) {
+			memccpy(parts[parts_count], start, '\0', (end - start));
+			parts_count++;
+
+			if (parts_count == PV_CTRL_MAX_SPLIT)
+				return parts_count;
+		}
+
+		start = end;
+	}
+
+	return parts_count;
+}
+
+bool pv_ctrl_utils_is_req_ok(struct evhttp_request *req, struct pv_ctrl_cb *cb)
+{
+	pv_log(DEBUG, "=== cb name: %s", cb->uri);
```

Comment:
in case you really use these === logs, we could also have something similar to event_rest, where we have very verbose logs under PV_LIBEVENT_DEBUG_MODE... Check _debug_log if you feel it would be worth it and maybe we could move it up to event/event_utils.h or something